File: battery-health.js
#!/usr/bin/env node
// Compatible runtimes: node, bun, deno

import { execSync } from "node:child_process";
import fs from "node:fs";
import path from "node:path";

/**
 * @file Battery health status analyzer tool.
 * @description Generates windows battery reports or reads inputs manually to print formatted health statistics with threshold coloring.
 */

// Colors utilizando secuencias ANSI
const CYAN = "\x1b[36m";
const WHITE = "\x1b[37m";
const GRAY = "\x1b[90m";
const YELLOW = "\x1b[33m";
const GREEN = "\x1b[32m";
const RED = "\x1b[31m";
const NC = "\x1b[0m";

/**
 * Prints the CLI usage manual instructions to the console.
 *
 * @returns {void}
 */
function showUsage() {
  console.log(`${CYAN}\n MODO DE USO:${NC}`);
  console.log(
    `${GRAY} ==========================================================================${NC}`,
  );
  console.log(`${WHITE} Modo Automático (Recomendado):${NC}`);
  console.log(`${YELLOW}   node batteryHealth.js --mode auto${NC}`);
  console.log(
    `${GRAY}   (Genera el reporte interno y extrae los datos de tu PC actual)\n${NC}`,
  );
  console.log(`${WHITE} Modo Manual:${NC}`);
  console.log(
    `${YELLOW}   node batteryHealth.js --mode manual --design 53015 --full 43243${NC}`,
  );
  console.log(
    `${GRAY} ==========================================================================\n${NC}`,
  );
}

// Mapeo simple de argumentos de terminal
const args = process.argv.slice(2);
const argMap = {};
for (let i = 0; i < args.length; i += 2) {
  argMap[args[i]] = args[i + 1];
}

const mode = argMap["--mode"] || "auto";
let designCapacity = parseFloat(argMap["--design"]) || 0;
let fullChargeCapacity = parseFloat(argMap["--full"]) || 0;

if (mode !== "auto" && mode !== "manual") {
  showUsage();
  process.exit(1);
}

/**
 * Evaluates the proportional remaining integrity of the battery.
 *
 * @param {number} fullCharge - Current maximum charge capacity in mWh.
 * @param {number} design - Original manufacturer capacity in mWh.
 * @returns {number} Percentage representation of overall battery health.
 */
function getBatteryHealth(fullCharge, design) {
  if (design === 0) return 0;
  return (fullCharge / design) * 100;
}

if (mode === "auto") {
  const tempDir = process.env.TEMP || process.env.TMP || "/tmp";
  const tempReport = path.join(tempDir, "batteryreport.html");

  try {
    // Generación silenciosa del reporte html
    execSync(`powercfg /batteryreport /output "${tempReport}"`, {
      stdio: "ignore",
    });

    if (fs.existsSync(tempReport)) {
      const html = fs.readFileSync(tempReport, "utf8");

      const designMatch = html.match(
        /DESIGN CAPACITY<\/span><\/td><td>([\d.,\s]+)mWh/,
      );
      if (designMatch) {
        designCapacity = parseFloat(designMatch[1].replace(/[\s.,]/g, ""));
      }

      const fullMatch = html.match(
        /FULL CHARGE CAPACITY<\/span><\/td><td>([\d.,\s]+)mWh/,
      );
      if (fullMatch) {
        fullChargeCapacity = parseFloat(fullMatch[1].replace(/[\s.,]/g, ""));
      }

      fs.unlinkSync(tempReport);
    }
  } catch (error) {
    // Falla controlada si el entorno no es compatible con powercfg
  }

  if (!designCapacity || !fullChargeCapacity) {
    console.error(
      "Error: No se pudieron obtener de forma automática las capacidades de la batería.",
    );
    process.exit(1);
  }
} else if (mode === "manual") {
  if (!designCapacity || !fullChargeCapacity) {
    console.error(
      "Error: En modo 'manual' debes proporcionar --design y --full.",
    );
    process.exit(1);
  }
}

const batteryHealth = getBatteryHealth(fullChargeCapacity, designCapacity);

let healthColor = RED;
if (batteryHealth >= 80) {
  healthColor = GREEN;
} else if (batteryHealth >= 50) {
  healthColor = YELLOW;
}

const line = "==================================";
console.log(`${CYAN}${line}${NC}`);
console.log(
  `${GRAY} DESIGN CAPACITY: ${YELLOW}${designCapacity}${GRAY}mWh${NC}`,
);
console.log(
  `${GRAY} FULL CHARGE CAPACITY: ${YELLOW}${fullChargeCapacity}${GRAY}mWh${NC}`,
);
console.log(
  `${GRAY} BATTERY HEALTH STATUS: ${healthColor}${batteryHealth.toFixed(2)}%${NC}`,
);
console.log(`${CYAN}${line}${NC}`);

File: battery_health.py
#!/usr/bin/env python3
import argparse
import os
import re
import subprocess
import sys

# Definición de códigos de color ANSI
CYAN = "\033[0;36m"
WHITE = "\033[1;37m"
GRAY = "\033[0;37m"
YELLOW = "\033[1;33m"
GREEN = "\033[0;32m"
RED = "\033[0;31m"
NC = "\033[0m"


def show_usage():
    """Prints terminal execution context manual helper."""
    print(f"{CYAN}\n MODO DE USO:{NC}")
    print(f"{GRAY} =========================================================================={NC}")
    print(f"{WHITE} Modo Automático (Recomendado):{NC}")
    print(f"{YELLOW}   python battery_health.py --mode auto{NC}")
    print(f"{GRAY}   (Genera el reporte interno y extrae los datos de tu PC actual)\n{NC}")
    print(f"{WHITE} Modo Manual:{NC}")
    print(f"{YELLOW}   python battery_health.py --mode manual --design 53015 --full 43243{NC}")
    print(f"{GRAY} ==========================================================================\n${NC}")


def get_battery_health(full_charge: float, design: float) -> float:
    """Calculates mathematical battery health ratio."""
    if design == 0:
        return 0.0
    return (full_charge / design) * 100


def main():
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument("--mode", default="auto", choices=["auto", "manual"])
    parser.add_argument("--design", type=float, default=0.0)
    parser.add_argument("--full", type=float, default=0.0)

    # Captura errores en argumentos para desviar a la guía personalizada
    try:
        args = parser.parse_args()
    except SystemExit:
        show_usage()
        sys.exit(1)

    design_capacity = args.design
    full_charge_capacity = args.full

    if args.mode == "auto":
        temp_dir = os.environ.get("TEMP") or os.environ.get("TMP") or "/tmp"
        temp_report = os.path.join(temp_dir, "batteryreport.html")

        try:
            # Ejecución controlada y redirección silenciosa de flujos
            subprocess.run(
                ["powercfg", "/batteryreport", "/output", temp_report],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                check=True
            )

            if os.path.exists(temp_report):
                with open(temp_report, "r", encoding="utf-8") as f:
                    html = f.read()

                design_match = re.search(r"DESIGN CAPACITY<\/span><\/td><td>([\d.,\s]+)mWh", html)
                if design_match:
                    design_capacity = float(re.sub(r"[\s.,]", "", design_match.group(1)))

                full_match = re.search(r"FULL CHARGE CAPACITY<\/span><\/td><td>([\d.,\s]+)mWh", html)
                if full_match:
                    full_charge_capacity = float(re.sub(r"[\s.,]", "", full_match.group(1)))

                os.remove(temp_report)
        except Exception:
            pass

        if not design_capacity or not full_charge_capacity:
            print("Error: No se pudieron obtener de forma automática las capacidades de la batería.", file=sys.stderr)
            sys.exit(1)

    elif args.mode == "manual":
        if not design_capacity or not full_charge_capacity:
            print("Error: En modo 'manual' debes proporcionar --design y --full.", file=sys.stderr)
            sys.exit(1)

    battery_health = get_battery_health(full_charge_capacity, design_capacity)

    # Evaluación condicional de salud de la celda de energía
    if battery_health >= 80:
        health_color = GREEN
    elif battery_health >= 50:
        health_color = YELLOW
    else:
        health_color = RED

    line = "=================================="
    print(f"{CYAN}{line}{NC}")
    print(f"{GRAY} DESIGN CAPACITY:       {YELLOW}{design_capacity:g}{GRAY} mWh{NC}")
    print(f"{GRAY} FULL CHARGE CAPACITY:  {YELLOW}{full_charge_capacity:g}{GRAY} mWh{NC}")
    print(f"{GRAY} BATTERY HEALTH STATUS: {health_color}{battery_health:.2f}%{NC}")
    print(f"{CYAN}{line}{NC}")


if __name__ == "__main__":
    main()
